refactor(allocator): move nested function out of `FixedSizeAllocatorPool::get`

overlookmotel opened the pull request:
Follow-on after #17023. Pure refactor.

Move code out of `FixedSizeAllocatorPool::get`, to make it easier to read. Also add comments.

camc314 reviewed (commented):
lol, claude suggested this change while I was working on it, and I deliberately didn't do it because it felt like I was mixing the two concerns of the pool and the fixed size allocator.

Iirc, it's also unsafe to call this function on a fixed size allocator if you didn't get it out of the pool?

overlookmotel commented:
> lol, claude suggested this change while I was working on it, and I deliberately didn't do it because it felt like I was mixing the two concerns of the pool and the fixed size allocator.

This whole thing with fixed size allocators is held together with gaffer tape and string. It's a disaster area, and UB is only a slip of the keyboard away. Everything is entwined with each-other, and it's gratuitously unsound. So I don't think separation of concerns really applies - and that's the problem - there isn't any!

> Iirc, it's also unsafe to call this function on a fixed size allocator if you didn't get it out of the pool?

Correct. But there's nowhere else to get one.

overlookmotel committed:
refactor(allocator): move nested function out of `FixedSizeAllocatorPool::get` (#17078)

Follow-on after #17023. Pure refactor.

Move code out of `FixedSizeAllocatorPool::get`, to make it easier to read. Also add comments.
